keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,694 @@
+type LocalQueuePendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Position indicates the position of the workload in the cluster queue.
+	Position *int32
+
+	// MaxPosition indicates the maximal position of the workload in the cluster
+	// queue returned in the head.
+	MaxPosition *int32
+}
+
+type LocalQueuePendingWorkloadsStatus struct {
+	// Head contains the list of top pending workloads.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Head []LocalQueuePendingWorkload
+
+	// LastChangeTime indicates the time of the last change of the structure.
+	LastChangeTime metav1.Time
+}
+
+// LocalQueueStatus defines the observed state of LocalQueue
+type LocalQueueStatus struct {
+...
+	// PendingWorkloadsStatus contains the information exposed about the current
+	// status of pending workloads in the local queue.
+	// +optional
+	PendingWorkloadsStatus *LocalQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Cluster Queue API
+
+```golang
+// ClusterQueuePendingWorkload contains the information identifying a pending workload
+// in the cluster queue.
+type ClusterQueuePendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Namespace indicates the name of the pending workload.
+	Namespace string
+}
+
+type ClusterQueuePendingWorkloadsStatus struct {
+	// Head contains the list of top pending workloads.
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=namespace
+	// +optional
+	Head []ClusterQueuePendingWorkload
+
+	// LastChangeTime indicates the time of the last change of the structure.
+	LastChangeTime metav1.Time
+}
+
+// ClusterQueueStatus defines the observed state of ClusterQueueStatus
+type ClusterQueueStatus struct {
+...
+	// PendingWorkloadsStatus contains the information exposed about the current
+	// status of the pending workloads in the cluster queue.
+	// +optional
+	PendingWorkloadsStatus *ClusterQueuePendingWorkloadsStatus
+...
+}
+```
+
+### Configuration
+
+```golang
+// Configuration is the Schema for the kueueconfigurations API
+type Configuration struct {
+...
+	// QueueVisibility is configuration to expose the information about the top
+	// pending workloads.
+	QueueVisibility *QueueVisibility
+}
+
+type QueueVisibility struct {
+	// LocalQueues is configuration to expose the information
+	// about the top pending workloads in the local queue.
+	LocalQueues *LocalQueueVisibility
+
+	// ClusterQueues is configuration to expose the information
+	// about the top pending workloads in the cluster queue.
+	ClusterQueues *ClusterQueueVisibility
+
+	// UpdateInterval specifies the time interval for updates to the structure
+	// of the top pending workloads in the queues.
+	// Defaults to 5s.
+	UpdateInterval time.Duration
+}
+
+type LocalQueueVisibility struct {
+	// MaxCount indicates the maximal number of pending workloads exposed in the
+	// local queue status. The maximal value is 4000.
+	// Defaults to 10.
+	MaxCount int32
```

Comment:
Right, when the value is 0, then we don't send the updates for the corresponding queue type. Further, when both values are 0, then the feature is disabled, and snapshot isn't taken.

Reply:
> I think it would be useful to temporarily stop updating the positions when the cluster performance is much degraded.

Yes, but then Kueue would need to be able to diagnose the cluster performance is bad. Typically, this is a job for P&F. 

Kueue on its side sets QPS limit to 20/s, but it wouldn't discriminate between updates for positions and other updates. An indirect way of controlling QPS for status updates is the `UpdateInterval` parameter. 

I think to deal with large scale we may need another solution, such as `On-demand http endpoint` via `Extension API Server`

Reply:
In general updates to ClusterQueue and LocalQueue are lower priority than updates to Workload or the jobs.

We could consider having a separate client for them.

Reply:
> Right, when the value is 0, then we don't send the updates for the corresponding queue type. Further, when both values are 0, then the feature is disabled, and snapshot isn't taken.

Thanks for the clarification :) SGTM

> Yes, but then Kueue would need to be able to diagnose the cluster performance is bad. Typically, this is a job for P&F.
> 
> Kueue on its side sets QPS limit to 20/s, but it wouldn't discriminate between updates for positions and other updates. An indirect way of controlling QPS for status updates is the UpdateInterval parameter.
> 
> I think to deal with large scale we may need another solution, such as On-demand http endpoint via Extension API Server

I agree with you that we should set appropriate parameters for the APF.

Reply:
> In general updates to ClusterQueue and LocalQueue are lower priority than updates to Workload or the jobs.
> 
> We could consider having a separate client for them.

Oh, it's a great idea!